Patterns: Handle empty results and 404 pages

ryelle opened the pull request:
Fixes #116. This adds the "empty state" and "404 page" styles. When a category or query result is empty, the page should say that, then display 6 recent patterns. If the page is not found, it should display the "404" page template. This template is handled by PHP, so we can remove the `bypass_404_page` filter.

The 404 page in the design also has "404 patterns" on it, but we don't have that category set up, so I've left it empty.

### Screenshots

Empty search result (Japanese site):

<img width="1026" alt="nothing-ja" src="https://user-images.githubusercontent.com/541093/125520557-d3cb2cd3-b164-4cbb-991c-9fb84b3cb61a.png">

A locale with no translated patterns (Korean) will also show the empty state on the homepage, but there are no recent patterns either. Ideally this would be a message to get involved, but that can happen in another PR/post launch.

<img width="998" alt="nothing-ko" src="https://user-images.githubusercontent.com/541093/125520558-8c8ba102-f720-40e0-a65b-2941d327f7e0.png">

The 404 page, for example, `wordpress.org/patterns/error404`:

<img width="1008" alt="404" src="https://user-images.githubusercontent.com/541093/125520551-c72b3342-8b25-4ae6-b1d2-16cac33e5fae.png">

### How to test the changes in this Pull Request:

1. Search for a string that doesn't exist
2. You should see the empty state
3. Go to a page or category that doesn't exist
4. You should see the 404 page

ryelle committed:
Pattern Grid: Rename loading state component.

This pulls the grid-logic out of the `PatternGrid` and allows the length of the skeleton to be customized to the query request.

ryelle committed:
Patterns: Show a notice state when Patterns has no content.

ryelle committed:
Add 404 page

StevenDufresne commented:
Looking at the first screenshot, this seems to work with the branch pre: aa775d8.

When I find nothing on the search page, I don't see my expected view:

![](https://d.pr/i/T6cqgd.png)

StevenDufresne commented:
Relates to #280.

ryelle commented:
Ah, because the search uses the separate component and I didn't fix that in the rebase. Will update this there too.

ryelle committed:
Add the empty state to "My Patterns" and search results

ryelle committed:
Show an empty state on favorites

This changes the initial state for `favorites` to `null`, so we can tell the difference between "not loaded" and "empty". Use this null value, in combination with the patterns query, to check if there are any favorite patterns - otherwise show the empty state.
